File: week7/identify_poet.pl
#!/usr/bin/perl

my %words;
foreach my $file (@ARGV) {
  open(F,"<$file") or die "$0: can't open $file\n";
  my @arr = <F>;
  @arr = split(/[^a-zA-Z]+/, join(' ',@arr));
  @arr = grep { $_ ne '' } @arr;
  foreach my $word (@arr) {
    $words{"$file"}{"$word"} = '';
  }
}

my %log_prob;
foreach my $file (glob "poems/*.txt") {
  my $artist = $file;
  $artist =~ s/poems\/([^\.]*).txt/\1/;
  $artist = join(' ', split(/_/,$artist));

  open(F,"<$file") or die "$0: can't open $file\n";
  my @arr = <F>;
  @arr = split(/[^a-zA-Z]+/, join(' ',@arr));
  @arr = grep { $_ ne '' } @arr;
  my $allCount = scalar @arr;

  foreach my $file (keys %words) {
    foreach my $word (keys %{ $words{"$file"} }) {

      if( exists $log_prob{"$artist"}{"$word"} ) {
        next;
      }
      my $count = 0;
      foreach my $elem (@arr) {
        chomp $elem;
        $elem = lc $elem;
        if ( $elem eq $word ) {
          $count ++;
        }
      }
      $log_prob{"$artist"}{"$word"} = log (($count + 1 ) / $allCount);
    }
  }
}

foreach $file (keys %words) {
  my %prob;
  foreach $artist (keys %log_prob) {
    $prob{"$artist"} = 0;
    foreach $word (keys %{ $words{"$file"} }) {
      $prob{"$artist"} += $log_prob{"$artist"}{"$word"};
    }
  }
  print "@{[%prob]}";

  # my $chosen = '';
  # foreach my $artist (keys %prob) {
  #   if($chosen eq '') {
  #     $chosen = $artist;
  #     next;
  #   }
  #   if($prob{"$artist"} > $prob{"$chosen"}) {
  #     $chosen = $artist;
  #   }
  # }

  # print "$chosen\n";
}

File: week7/identify_poet.py
#!/usr/bin/python

import sys, re, glob, math

debug = False
files = sys.argv[1:]
if(sys.argv[1] == "-d") :
    debug = True
    files = sys.argv[2:]

words = {}
for fil in files :
    F = open(fil,"r")
    arr = list(filter(None, re.split(r'[^a-zA-Z]+',' '.join(F.readlines()))))
    words[fil] = {}
    for word in arr :
        word = word.strip().lower()
        words[fil][word] = words[fil].get(word,1) + 1

log_prob = {}
for fil in sorted(glob.glob("poems/*.txt")) :
    artist = re.sub(r'poems\/([^\.]*).txt',r'\1',fil)
    artist = ' '.join(artist.split('_'))

    F = open(fil,"r")
    arr = list(filter(None, re.split(r'[^a-zA-Z]+',' '.join(F.readlines()))))
    allCount = len(arr)

    log_prob[artist] = {}
    for fil1, d in words.items() :
        for word in d :
            if (word in log_prob[artist]) :
                continue

            count = len(list(filter(lambda x: x == word,list(map(lambda x: x.lower(), arr)))))
            log_prob[artist][word] = math.log((count+1) / allCount)
    F.close()

for fil in sorted(words.keys()) :
    prob = {}
    for artist, d in log_prob.items():
        prob[artist] = 0
        for word in d :
            prob[artist] += ( log_prob[artist][word] * d[word])

    chosen = ''
    sortedKeys = sorted(prob, key=prob.get)
    if(debug) :
        for artist in sortedKeys :
            print ("%s: log_probability of %.1f for %s"% (fil,prob[artist],artist))
    print("%s most resembles the work of %s (log-probability=%.1f)" % (fil,sortedKeys[0], prob[sortedKeys[0]]))
